wit/types.wit

```diff
@@ -16,29 +25,29 @@ interface messaging-types {
     /// The interface doesn't highlight this difference in the type itself as that's uniquely a consumer issue.
     type channel = string;
   
-    /// Configuration includes a required list of channels the guest is subscribing to, and an optional list of extensions key-value pairs 
-    /// (e.g., partitions/offsets to read from in Kafka/EventHubs, QoS etc.).
+    /// Configuration includes a required list of channels the guest is subscribing to, and an
+    /// optional list of extensions key-value pairs (e.g., partitions/offsets to read from in 
+    /// Kafka/EventHubs, QoS etc.).
     record guest-configuration {
         channels: list<channel>,
         extensions: option<list<tuple<string, string>>>
```

Comment:
It seems like if wasi-messaging has `extensions` that are specially interpreted in important (i.e., not just performance-hinting) ways, then it will produce components that only run correctly on one implementation (say Kafka), which prevents portability.  Could we instead have the motivating use cases (say, partitions/offsets) be explicitly defined by wasi-messaging as separate `interface`s so that a component can declaratively say that it needs those features and different host messaging systems can implement them or not (and if not, maybe they can be virtualized via component).

Reply:
The fact that messaging is used for so many different applications was one of the main pain points I ran into when writing this interface - that is, different possible implementors like: Service Bus, Event Hubs, Kafka, NATS, etc. are all used for different purposes and, with that, have different strengths (e.g., some are used for load balancing, others for decoupling microservices, others for stream processing, and some for telemetry/logging just to name a few applications). For a specific example: Event Hubs does not have the capability to read messages from a consumer group, only send; while, on the other hand, Service Bus has advanced functions to not only read messages, but also peek and so on. With that, my main fear with breaking this interface into smaller pieces like so is that we will just end up with a close-to-zero common denominator and just have interfaces that pretty much represent a service offering (e.g., a “Kafka“ interface). Anyway, a good starting point could be to break messaging into its main applications and use that to bucket different implementors - this way, we can compare their APIs and use that as a blueprint to better type smt like `extensions` and see if that gets us to a point where this interface is sufficiently portable.

Reply:
Also - I had sort of started doing this in here: https://github.com/WebAssembly/wasi-messaging/pull/9#discussion_r1178061562 (i.e., splitting the interface for different applications - e.g., load balancer, but the change got overwritten).

Reply:
Yeah, I can totally imagine there is a wide variety of functionality here.  In the limit, every single messaging implementation could define its own WIT interface (e.g., `apache:kafka`, `azure:service-bus`, etc), and this isn't a bad idea: a customer may wish to use the full feature set of a particular implementation and so an implementation-specific WIT is appropriate (defined and owned by that implementation, not WASI).  Given this option as a baseline, you could say that the role of `wasi:messaging` is to handle cases where a customer intentionally wants portability over multiple implementations (if they didn't, they could just use the WIT for the impl they wanted, and have an altogether better DX).

Given that framing, where the primary value we're adding by defining `wasi:messaging` is abstracting over different implementations, we can (1) provide a least-common-denominator (that works portably over all implementations, but necessarily leaves a bunch of functionality on the table) or (2) start with a least-common-denominator, and add different chunks of optional functionality that are only supported by a (>1-sized) subset of implementations.  (Or start with (1) and then move on to (2).)  I think the real "hard work" of standardization is right-sizing the chunks so that we end up with far fewer chunks than implementations (otherwise, might as well use the impl-specific WITs).

Reply:
That's good framing - I am aligned with you on providing a least-common-denominator interface that can be built upon. And, I suppose we're sort of on the way w/ (2) already by w/ implementing `request-reply` as an optional interface w/ the supporting rationale that it is a common pattern across a wide variety of messaging implementors albeit not supported by all.

Reply:
Great!  And what's neat about having `request-reply` as an explicitly-declared separate interface is that if you want to run your component on a messaging impl without native `request-reply`, you might be able to virtualize (in the host or in an adapter component) in terms of the underlying primitives that are present.

So then do you think we can remove `extensions`?

Reply:
I'd still leave `extensions` for things like perfomance hinting and alike  - but I do think we can start making an effort to move some of the things that are important and are currently intended for `extensions` (e.g., partition offsets and whatnot) out into their own optional interfaces. @thomastaylor312 , do you want to do this as part of this PR? If not, I can work on it on a follow-up too 👍

Reply:
I wonder if even performance hints can benefit, portability-wise, by having them explicit named optional fields (of either a record or a resource type (like we did in wasi-http with [`request-options`](https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L331)) so that the meaning of the hint is clear (even if it is ignored).  Or at least this is a conservative way to start, and if we find that there are just too many, then maybe we reconsider.  As a general learning from standards work (many times over on the Web), if you create an open-ended "stick your implementation-specific stuff here" hole, it will end up being used in a way that breaks portability, so it's best to start without it and hold off as long as possible.

Reply:
Yeah I can move that stuff out here @danbugs!

Reply:
Fair play - I'm just not sure how we should balance flexibility and portability in this scenario. I feel like a lot of users might want something like `extensions`, but I also can't disagree that it inevitably hurts portability...

Regardless, I suppose starting with removing it and then adding back functionality per user requests could be the best way to go.

Reply:
So I currently have left extensions in. Given that they are optional and non-structured, I don't think it hurts per se to have some guest hinting something to Kafka, even though they're connected to Event Hubs. Still ok to take it out if we worry about the portability aspect

Reply:
Are there any specific high-value examples of hints you can point to which would make the first release of this interface non-viable?  If so, again, maybe they could be optional fields of a record/resource; if not, then perhaps, just based on the general 80/20-rule principle that you've been using, we should leave if off for now (we can always add, but much harder to subtract).

Reply:
I personally don't have any, but I also don't use Kafka a bunch, which is what I think has been stated here. @danbugs and @devigned do either of you have any strong examples here? Otherwise I'm inclined to agree with @lukewagner and take it out

Reply:
I'm good w/ taking it out 👍